src/map/clan.c

```diff
@@ -0,0 +1,839 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#define HERCULES_CORE
+
+#include "config/core.h"
+#include "clan.h"
+
+#include "map/clif.h"
+#include "map/chrif.h"
+#include "map/homunculus.h"
+#include "map/log.h"
+#include "map/mercenary.h"
+#include "map/mob.h"
+#include "map/npc.h"
+#include "map/pc.h"
+#include "map/pet.h"
+#include "map/script.h"
+#include "map/skill.h"
+#include "map/status.h"
+#include "common/HPM.h"
+#include "common/conf.h"
+#include "common/cbasetypes.h"
+#include "common/db.h"
+#include "common/memmgr.h"
+#include "common/mapindex.h"
+#include "common/nullpo.h"
+#include "common/showmsg.h"
+#include "common/socket.h"
+#include "common/strlib.h"
+#include "common/timer.h"
+#include "common/utils.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+struct clan_interface clan_s;
+struct clan_interface *clan;
+
+/**
+ * Searches a Clan by clan_id
+ *
+ * @param clan_id Clan ID
+ * @return struct clan*
+ */
+struct clan *clan_search(int clan_id)
+{
+	if (clan_id <= 0) {
+		return NULL;
+	}
+	return (struct clan *)idb_get(clan->db, clan_id);
+}
+
+/**
+ * Searches a Clan by clan_name
+ *
+ * @param name Clan Name
+ * @return struct clan*
+ */
+struct clan *clan_searchname(const char *name)
+{
+	struct clan *c;
+	struct DBIterator *iter;
+
+	nullpo_retr(NULL, name);
+
+	iter = db_iterator(clan->db);
+	for (c = dbi_first(iter); dbi_exists(iter); c = dbi_next(iter)) {
+		if (strncmpi(c->name, name, NAME_LENGTH) == 0) {
+			break;
+		}
+	}
+	dbi_destroy(iter);
+	return c;
+}
+
+/**
+ * Returns the first online character of clan
+ * @method clan_getonlinesd
+ * @param  c                clan structure
+ * @return                  map_session_data
+ */
+struct map_session_data *clan_getonlinesd(struct clan *c)
+{
+	int i;
+	nullpo_retr(NULL, c);
+
+	ARR_FIND(0, c->max_member, i, c->member[i].sd != NULL);
+	return (i < c->max_member) ? c->member[i].sd : NULL;
+}
+
+/**
+ * Returns the member index of given Player
+ *
+ * @param c Clan Data
+ * @param account_id Player's Account ID
+ * @param char_id Player's Char ID
+ * @return int
+ */
+int clan_getindex(const struct clan *c, int account_id, int char_id)
+{
+	int i;
+	nullpo_retr(INDEX_NOT_FOUND, c);
+
+	ARR_FIND(0, c->max_member, i, c->member[i].account_id == account_id && c->member[i].char_id == char_id);
+
+	if (i == c->max_member) {
+		return INDEX_NOT_FOUND;
+	}
+	return i;
+}
+
+/**
+ * Starts clan buff
+ */
+void clan_buff_start(struct map_session_data *sd, struct clan *c)
+{
+	nullpo_retv(sd);
+	nullpo_retv(c);
+	if (c->buff.icon <= 0) {
+		return;
+	}
+	clif->sc_load(&sd->bl, sd->bl.id, SELF, c->buff.icon, 0, c->clan_id, 0);
+	script->run(c->buff.script, 0, sd->bl.id, npc->fake_nd->bl.id);
```

Comment:
There is no prior check, even by the callers, for a null script. This could be bad...
The check should ideally be within this function to silently return if there's nothing to buff with. (Or send the icon even if there's no script, whichever you prefer).